src/spin_mutex.cc

```diff
@@ -0,0 +1,34 @@
+#include "cris/core/spin_mutex.h"
+
+#include <cstddef>
+#include <thread>
+
+namespace cris::core {
+
+void HybridSpinMutex::lock() {
+    // 50k loop ~= 500k x86 inst ~= 1M cycles ~= 500us@2GHz
+    // To reduce contention, 100 loop ~= 1k x86 inst ~= 2k cycles ~= 1us@2GHz between atomic variable checks.
```

Comment:
```
$ cat 1.cc && clang++-12 1.cc -std=c++20 -Wall -Wextra -o 1 && for i in {1..10}; do ./1; done
#include <iostream>
#include <x86intrin.h>

using namespace std;

int main() {
    auto st = __rdtsc();
    for (int i = 0; i < 100; ++i) {}
    auto en = __rdtsc();

    cout << (en - st) << endl;
    return 0;
}
714
674
598
618
690
514
482
526
496
580
```

@xkszltl I would say we overestimate the cycles a little bit.

Reply:
My TSC Freq is about 2.4941 GHz, FYI

Reply:
1. This is opt-dependent, please test with -O3.
2. Your avg at 590, so calibrate it up by 4x.

You may also need some extras inst under O3.

```
% clang++ orz.cpp -std=c++20 -o orz && for i in $(seq 10); do ./orz; done    
410
462
520
478
494
464
480
470
512
488
% clang++ orz.cpp -std=c++20 -o orz -O3 && for i in $(seq 10); do ./orz; done
18
28
18
20
18
22
20
20
22
20
```